packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -177,3 +204,149 @@ class DocumentBoundary extends TextBoundary {
+    }
+    int index = position.offset;
+    if (position.affinity == TextAffinity.downstream && !TextLayoutMetrics.isWhitespace(_text.codeUnitAt(index))) {
+      return position;
+    }
+
+    while ((index -= 1) >= 0) {
+      if (!TextLayoutMetrics.isWhitespace(_text.codeUnitAt(index))) {
+        return TextPosition(offset: index + 1, affinity: TextAffinity.upstream);
+      }
+    }
+    return const TextPosition(offset: 0);
+  }
+
+  @override
+  TextPosition getTrailingTextBoundaryAt(TextPosition position) {
+    // Handles outside of right bound.
+    if (position.offset >= _text.length) {
+      return TextPosition(offset: _text.length, affinity: TextAffinity.upstream);
+    }
+    // Handles outside of left bound.
+    if (position.offset < 0 || (position.offset == 0 && position.affinity == TextAffinity.upstream)) {
+      position = const TextPosition(offset: 0);
+    }
+
+    int index = position.offset;
+    if (position.affinity == TextAffinity.upstream && !TextLayoutMetrics.isWhitespace(_text.codeUnitAt(index - 1))) {
+      return position;
+    }
+
+    for (; index < _text.length; index += 1) {
+      if (!TextLayoutMetrics.isWhitespace(_text.codeUnitAt(index))) {
+        return TextPosition(offset: index);
+      }
+    }
+    return TextPosition(offset: _text.length, affinity: TextAffinity.upstream);
+  }
+}
+
+/// Gets the boundary by calling the [outer] and pipe the result to
+/// [inner].
+class _ExpandedTextBoundary extends TextBoundary {
+  /// Creates a [_ExpandedTextBoundary] with inner and outter boundaries
+  const _ExpandedTextBoundary({required this.inner, required this.outer});
+
+  /// The inner boundary to call with the result from [outer].
+  final TextBoundary inner;
+
+  /// The outer boundary to call with the input position.
+  ///
+  /// The result is piped to the [inner] before returning to the caller.
+  final TextBoundary outer;
+
+  @override
+  TextPosition getLeadingTextBoundaryAt(TextPosition position) {
+    return inner.getLeadingTextBoundaryAt(
+      outer.getLeadingTextBoundaryAt(position),
+    );
+  }
+
+  @override
+  TextPosition getTrailingTextBoundaryAt(TextPosition position) {
+    return inner.getTrailingTextBoundaryAt(
+      outer.getTrailingTextBoundaryAt(position),
+    );
+  }
+}
+
+/// A proxy text boundary that will push input text position forward or backward
+/// one affinity unit before sending it to the [textBoundary].
+///
+/// If the [forward] is true, this proxy text boundary push the position
+/// forward; otherwise, backward.
+///
+/// To push a text position forward one affinity unit, this proxy converts
+/// affinity to downstream if it is upstream; otherwise it increase the offset
+/// by one with its affinity sets to upstream. For example,
+/// `TextPosition(1, upstream)` becomes `TextPosition(1, downstream)`,
+/// `TextPosition(4, downstream)` becomes `TextPosition(5, upstream)`.
+class PushTextPosition extends TextBoundary {
```

Comment:
Would it be better to use TextBoundary composition instead of taking a `TextBoundary` explicitly? With the former say we create 2 constants `PreviousCaretLocation` and `NextCaretLocation`, 

So today's 
```
_PushTextPosition(mixedBoundary, intent.forward);
``` 
will be equivalent to 

```dart
(intent.forward ? TextBoundary.NextCaretLocation : TextBoundary.PreviousCaretLocation) + mixedBoundary;
```
or 
```dart
_PushTextPosition(info.forward) + mixedBoundary;
```

The order of operations will be clearer (from left to right, with _PushTextPosition it's not immediately clear if mixedBoundary will be applied first or push will be applied first).

Reply:
I find it hard to understand than the current code

PushTextPosition( WhiteSpaceBoundary() + WordBoundary(), true)
will become 
(WordBoundary()+ WordBoundary).nextCaretLocation where the nextCaretLocation is from right to left.

Reply:
I'm not suggesting making that a getter. Your example will be equivalent to `NextCaretLocation + WhiteSpaceBoundary + WordBoundary`